util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,284 @@
+// and collecting the metric output to be returned to the controller when requested.
+package worker
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/perf-tests/util-images/network/netperfbenchmark/pkg/util"
+
+	"k8s.io/klog"
+)
+
+// Worker hold data required for facilitating measurement
+type Worker struct {
+	resultStatus chan string
+	result       []string
+	tstDuration  string
+	tstProtocol  string
+	startedAt    int64
+	futureTime   int64
+}
+
+// startResponse is the response sent back to controller on starting a measurement
+type startWorkResponse struct {
+	error string
+}
+
+// metricResponse is the response sent back to controller after collecting measurement
+type metricResponse struct {
+	Result          []float64
+	WorkerStartTime string
+	Error           string
+}
+
+// http  listen ports
+const (
+	workerListenPort = "5003"
+	httpPort         = "5301"
+)
+
+var (
+	//Command arguments for each protocol.This supports templates("{}"),the value in template will be replaced by value in http request
+	// Iperf command args:
+	// -c <destIP> : connect to destIP(as client)
+	// -f K : report format KBytes/sec
+	// -l 20 : read/write buffer size 20 bytes
+	// -b 1M : bandwidth 1 Mbits/sec
+	// -i 1 : report stats every 1 sec
+	// -t duration : run  <duration> seconds
+	// -u : for udp measurement
+	// -e : enhanced reports, gives more metrics for udp
+	// -s : run in server mode
+	// -P numCls: handle <numCls> number of clients before disconnecting
+	udpClientCmd = []string{"-c", "{destIP}", "-u", "-f", "K", "-l", "20", "-b", "1M", "-e", "-i", "1", "-t", "{duration}"}
+	udpServerCmd = []string{"-s", "-f", "K", "-u", "-e", "-i", "{duration}", "-P", "{numCls}"}
+	tcpServerCmd = []string{"-s", "-f", "K", "-i", "{duration}", "-P", "{numCls}"}
+	tcpClientCmd = []string{"-c", "{destIP}", "-f", "K", "-l", "20", "-b", "1M", "-i", "1", "-t", "{duration}"}
+	// Siege command args:
+	// -d1 : random delay between 0 to 1 sec
+	// -t<duration>S : run test for <duration> seconds
+	// -c1 : one concurrent user
+	httpClientCmd = []string{"http://" + "{destIP}" + ":" + httpPort + "/test", "-d1", "-t" + "{duration}" + "S", "-c1"}
+)
+
+// Start worker
+func (w *Worker) Start() {
+	w.resultStatus = make(chan string)
+	w.listenToController()
+}
+
+func (w *Worker) listenToController() {
+	h := map[string]func(http.ResponseWriter, *http.Request){"/startTCPServer": w.StartTCPServer,
+		"/startTCPClient": w.StartTCPClient, "/startUDPServer": w.StartUDPServer, "/startUDPClient": w.StartUDPClient,
+		"/startHTTPServer": w.StartHTTPServer, "/startHTTPClient": w.StartHTTPClient, "/metrics": w.Metrics}
+	w.startListening(workerListenPort, h)
+	klog.Info("Started listening to Server")
+}
+
+func (w *Worker) startListening(port string, handlers map[string]func(http.ResponseWriter, *http.Request)) {
+	klog.Info("About to listen to controller requests")
+	for p, h := range handlers {
```

Comment:
for port, handler := ...

Reply:
Modified. have made it as "urlPath, handle:=" as by 'p' I meant path